### PR TITLE
[FLIRT] Enforce arch match on sig, zfs output n of matches and added pac support & tests

### PR DESCRIPTION
**Pull request by wargio**
Tests

 

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

- Support `.pac` FLIRT format
- Output number of matches when `zfs` is called
- Enforces the arch type when calling `zfs` with a `.sig` FLIRT file

**Test plan**

Green tests

**Closing issues**

Related to #817 request

### CODE REVIEW

File: librz/core/csign.c [hunk 1]
```diff
@@ -95,15 +95,24 @@ static void flirt_print_node(const RzFlirtNode *node, int indent) {
 RZ_API void rz_core_flirt_dump(RZ_NONNULL const char *flirt_file) {
 	rz_return_if_fail(RZ_STR_ISNOTEMPTY(flirt_file));
 
+	const char *extension = rz_str_lchr(flirt_file, '.');
+	if (RZ_STR_ISEMPTY(extension) || (strcmp(extension, ".sig") != 0 && strcmp(extension, ".pac") != 0)) {
```
Comment (XVilka) [inline librz/core/csign.c:99]: I think it's better to determine by the file magic (if there is any) than by extension.

File: librz/include/rz_flirt.h [hunk 2]
```diff
@@ -15,6 +15,69 @@ extern "C" {
...
+#define RZ_FLIRT_SIG_ARCH_ALPHA     24 // DEC Alpha
+#define RZ_FLIRT_SIG_ARCH_HPPA      25 // Hewlett-Packard PA-RISC
+#define RZ_FLIRT_SIG_ARCH_H8500     26 // Hitachi H8/500
+#define RZ_FLIRT_SIG_ARCH_TRICORE   27 // Tasking Tricore
+#define RZ_FLIRT_SIG_ARCH_DSP56K    28 // Motorola DSP5600x
+#define RZ_FLIRT_SIG_ARCH_C166      29 // Siemens C166 family
+#define RZ_FLIRT_SIG_ARCH_ST20      30 // SGS-Thomson ST20
+#define RZ_FLIRT_SIG_ARCH_IA64      31 // Intel Itanium IA64
+#define RZ_FLIRT_SIG_ARCH_I960      32 // Intel 960
+#define RZ_FLIRT_SIG_ARCH_F2MC      33 // Fujistu F2MC-16
```
Comment (XVilka) [inline librz/include/rz_flirt.h]: Fujitsu

File: librz/include/rz_flirt.h [hunk 3]
```diff
@@ -15,6 +15,69 @@ extern "C" {
...
+#define RZ_FLIRT_SIG_ARCH_SH        18 // Renesas (formerly Hitachi) SuperH
+#define RZ_FLIRT_SIG_ARCH_NET       19 // Microsoft Visual Studio.Net
+#define RZ_FLIRT_SIG_ARCH_AVR       20 // Atmel 8-bit RISC processor(s)
+#define RZ_FLIRT_SIG_ARCH_H8        21 // Hitachi H8/300, H8/2000
+#define RZ_FLIRT_SIG_ARCH_PIC       22 // Microchip's PIC
+#define RZ_FLIRT_SIG_ARCH_SPARC     23 // SPARC
+#define RZ_FLIRT_SIG_ARCH_ALPHA     24 // DEC Alpha
+#define RZ_FLIRT_SIG_ARCH_HPPA      25 // Hewlett-Packard PA-RISC
+#define RZ_FLIRT_SIG_ARCH_H8500     26 // Hitachi H8/500
+#define RZ_FLIRT_SIG_ARCH_TRICORE   27 // Tasking Tricore
```
Comment (XVilka) [inline librz/include/rz_flirt.h]: What is `Tasking`? Just Tricore, I guess.

File: librz/include/rz_flirt.h [hunk 4]
```diff
@@ -15,6 +15,69 @@ extern "C" {
...
+#define RZ_FLIRT_SIG_ARCH_MIPS      12 // MIPS
+#define RZ_FLIRT_SIG_ARCH_ARM       13 // Advanced RISC Machines
+#define RZ_FLIRT_SIG_ARCH_TMSC6     14 // Texas Instruments TMS320C6x
+#define RZ_FLIRT_SIG_ARCH_PPC       15 // PowerPC
+#define RZ_FLIRT_SIG_ARCH_80196     16 // Intel 80196
+#define RZ_FLIRT_SIG_ARCH_Z8        17 // Z8
+#define RZ_FLIRT_SIG_ARCH_SH        18 // Renesas (formerly Hitachi) SuperH
+#define RZ_FLIRT_SIG_ARCH_NET       19 // Microsoft Visual Studio.Net
+#define RZ_FLIRT_SIG_ARCH_AVR       20 // Atmel 8-bit RISC processor(s)
+#define RZ_FLIRT_SIG_ARCH_H8        21 // Hitachi H8/300, H8/2000
```
Comment (XVilka) [inline librz/include/rz_flirt.h:39]: Then it's also `Renesas (formerly Hitachi)`

File: librz/signature/flirt.c [hunk 5]
```diff
@@ -388,15 +351,28 @@ RZ_API void rz_sign_flirt_node_free(RZ_NULLABLE RzFlirtNode *node) {
...
+	free(node->pattern_mask);
 	free(node->pattern_bytes);
 	rz_list_free(node->module_list);
 	rz_list_free(node->child_list);
 	free(node);
 }
 
-static int module_match_buffer(RzAnalysis *analysis, const RzFlirtModule *module,
-	ut8 *b, ut64 address, ut32 buf_size) {
+/* Returns true if b matches the pattern in node. */
```
Comment (XVilka) [inline librz/signature/flirt.c]: Convert into Doxygen then maybe?

File: librz/signature/flirt.c [hunk 6]
```diff
@@ -892,7 +851,7 @@ static ut8 read_node_variant_mask(RzFlirtNode *node, ParseStatus *b) {
 }
 
 static bool read_node_bytes(RzFlirtNode *node, ParseStatus *b) {
-	/*Reads the node bytes, and also sets the variant bytes in variant_bool_array*/
+	/*Reads the node bytes, and also sets the variant bytes in pattern_mask*/
```
Comment (XVilka) [inline librz/signature/flirt.c]: Spaces.

File: librz/signature/pac.c [hunk 7]
```diff
@@ -0,0 +1,369 @@
+// SPDX-FileCopyrightText: 2021 RizinOrg <info@rizin.re>
+// SPDX-FileCopyrightText: 2021 deroad <wargio@libero.it>
+// SPDX-License-Identifier: LGPL-3.0-only
```
Comment (XVilka) [inline librz/signature/pac.c:3]: Can you please add a short explanation of what pac is and a rough file format structure if any? Using Doxygen.

File: librz/signature/pac.c [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+
+	//eprintf("pattern: ");
+	for (ut32 i = 0; i < n_bytes; ++i) {
+		ut32 p = i << 1;
+		char high = in_pattern[p];
+		char low = in_pattern[p + 1];
+		if (IS_HEXCHAR(high) && IS_HEXCHAR(low)) {
+			bytes[i] = parse_byte(high, low);
+			mask[i] = 0;
+			//eprintf("%02X(%c%c) ", bytes[i], high, low);
```
Comment (XVilka) [inline librz/signature/pac.c]: Remove those

File: librz/signature/pac.c [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+	free(mask);
+	free(bytes);
+	return false;
+}
+
+/**
+ * Expects one of these line formats and lines may end with '\r'
+ * 
+ * 4154554889FD534889F3C60700E8........C6441DFF004189C485C07515BE2E 07 FAEE 003B :0000@ Curl_gethostname ^000E gethostname ^0027 strchr ........4885C07403C600004489E05B5D415CC3
+ * 31C04885D2741F488D4417FF4839C77610EB1D0F1F4400004883E8014839C777 13 9867 0033 :0000 Curl_memrchr
```
Comment (XVilka) [inline librz/signature/pac.c:127]: Split the long line?

File: librz/signature/pac.c [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+
+/**
+ * \brief Parses the RzBuffer containing a FLIRT signature in string format and returns an RzFlirtNode
+ * 
+ * \param  flirt_buf     The buffer to read
+ * \return               Parsed FLIRT node
+ */
+RZ_API RZ_OWN RzFlirtNode *rz_sign_flirt_parse_string_buffer(RZ_NONNULL RzBuffer *flirt_buf) {
+	rz_return_val_if_fail(flirt_buf, NULL);
+	char buffer[1024];
```
Comment (XVilka) [inline librz/signature/pac.c]: Make it a define maybe?

**Review by XVilka — CHANGES_REQUESTED** [review]
Very good job and a clean code. Mostly stylistic picks.

File: librz/core/csign.c (hunk 1 above)
Comment (wargio) [reply librz/core/csign.c:99]: pac does not have any magic

File: librz/signature/pac.c (hunk 9 above)
Comment (wargio) [reply librz/signature/pac.c:127]: no, that is an example of line

File: librz/core/csign.c (hunk 1 above)
Comment (wargio) [reply librz/core/csign.c:99]: also for compressed flirts, there is a magic check

**Review by XVilka — APPROVED** [review]

**Pull request merged**